File: ios/ReactNativeArView.swift
import ExpoModulesCore
import ARKit
import SceneKit
import GLTFKit2

class ReactNativeArView: ExpoView, ARSCNViewDelegate, ARSessionDelegate, UIGestureRecognizerDelegate {

    // MARK: - Static instance for module access

    static weak var currentInstance: ReactNativeArView?

    // MARK: - Events

    let onTrackingStateChange = EventDispatcher()
    let onPlaneDetected = EventDispatcher()
    let onModelLoaded = EventDispatcher()
    let onModelPlaced = EventDispatcher()
    let onModelError = EventDispatcher()
    let onSceneChange = EventDispatcher()
    let onARError = EventDispatcher()

    // MARK: - AR View

    private let arSceneView = ARSCNView()

    // MARK: - Scene management

    private var sceneStack: [ARSceneState] = []
    private var currentModels: [ModelConfig] = []
    private var placedAnchors: [AnchorNodePair] = []

    // MARK: - Pending models for tap-to-place

    private var pendingModelConfigs: [ModelConfig] = []
    private var currentPendingIndex: Int = 0

    // MARK: - Node-to-config map for gesture settings

    private var nodeConfigMap: [SCNNode: ModelConfig] = [:]

    // MARK: - Currently selected node for gestures

    private var activeGestureNode: SCNNode?

    // MARK: - Tracking state

    private var lastTrackingState: String = "unavailable"
    private var detectedPlaneIds: Set<String> = []

    // MARK: - Layout flag

    private var hasStartedSession = false

    // MARK: - Plane overlay nodes

    private var planeOverlayNodes: [ARAnchor: SCNNode] = [:]
    private var hasPlacedFirstModel = false

    // MARK: - Init

    required init(appContext: AppContext? = nil) {
        super.init(appContext: appContext)
        clipsToBounds = true

        ReactNativeArView.currentInstance = self

        // Configure ARSCNView
        arSceneView.delegate = self
        arSceneView.session.delegate = self
        arSceneView.autoenablesDefaultLighting = true
        arSceneView.automaticallyUpdatesLighting = true
        arSceneView.translatesAutoresizingMaskIntoConstraints = false

        addSubview(arSceneView)

        // Set up gesture recognizers
        let tapGesture = UITapGestureRecognizer(target: self, action: #selector(handleTap(_:)))
        arSceneView.addGestureRecognizer(tapGesture)

        let pinchGesture = UIPinchGestureRecognizer(target: self, action: #selector(handlePinch(_:)))
        pinchGesture.delegate = self
        arSceneView.addGestureRecognizer(pinchGesture)

        let rotationGesture = UIRotationGestureRecognizer(target: self, action: #selector(handleRotation(_:)))
        rotationGesture.delegate = self
        arSceneView.addGestureRecognizer(rotationGesture)

        let panGesture = UIPanGestureRecognizer(target: self, action: #selector(handlePan(_:)))
        panGesture.delegate = self
        arSceneView.addGestureRecognizer(panGesture)
    }

    // Allow simultaneous gesture recognition (pinch + rotation)
    func gestureRecognizer(
        _ gestureRecognizer: UIGestureRecognizer,
        shouldRecognizeSimultaneouslyWith otherGestureRecognizer: UIGestureRecognizer
    ) -> Bool {
        return true
    }

    // MARK: - Layout & Lifecycle

    override func layoutSubviews() {
        super.layoutSubviews()
        arSceneView.frame = bounds

        if !hasStartedSession && bounds.width > 0 && bounds.height > 0 {
            hasStartedSession = true
            startARSession()
        }
    }

    private func startARSession() {
        guard ARWorldTrackingConfiguration.isSupported else {
            onARError([
                "code": "AR_NOT_SUPPORTED",
                "message": "ARWorldTrackingConfiguration is not supported on this device"
            ])
            return
        }

        let configuration = ARWorldTrackingConfiguration()
        configuration.planeDetection = [.horizontal, .vertical]
        configuration.environmentTexturing = .automatic

        // Enable scene depth if LiDAR is available
        if ARWorldTrackingConfiguration.supportsFrameSemantics(.sceneDepth) {
            configuration.frameSemantics.insert(.sceneDepth)
        }

        arSceneView.session.run(configuration, options: [.resetTracking, .removeExistingAnchors])
    }

    override func removeFromSuperview() {
        arSceneView.session.pause()
        if ReactNativeArView.currentInstance === self {
            ReactNativeArView.currentInstance = nil
        }
        super.removeFromSuperview()
    }

    // MARK: - ARSCNViewDelegate (Plane Visualization)

    func renderer(_ renderer: SCNSceneRenderer, didAdd node: SCNNode, for anchor: ARAnchor) {
        guard let planeAnchor = anchor as? ARPlaneAnchor else { return }

        // Create translucent plane overlay
        let plane = SCNPlane(
            width: CGFloat(planeAnchor.extent.x),
            height: CGFloat(planeAnchor.extent.z)
        )
        let material = SCNMaterial()
        material.diffuse.contents = UIColor(white: 1.0, alpha: 0.3)
        material.isDoubleSided = true
        plane.materials = [material]

        let planeNode = SCNNode(geometry: plane)
        // SCNPlane is vertical by default, rotate to be horizontal
        planeNode.eulerAngles.x = -.pi / 2
        planeNode.isHidden = hasPlacedFirstModel

        node.addChildNode(planeNode)
        planeOverlayNodes[anchor] = planeNode

        // Fire plane detected event (dedup by plane ID)
        let planeId = anchor.identifier.uuidString
        if !detectedPlaneIds.contains(planeId) {
            detectedPlaneIds.insert(planeId)
            let type: String = planeAnchor.alignment == .vertical ? "vertical" : "horizontal"
            DispatchQueue.main.async { [weak self] in
                self?.onPlaneDetected(["id": planeId, "type": type])
            }
        }
    }

    func renderer(_ renderer: SCNSceneRenderer, didUpdate node: SCNNode, for anchor: ARAnchor) {
        guard let planeAnchor = anchor as? ARPlaneAnchor,
              let planeNode = planeOverlayNodes[anchor],
              let plane = planeNode.geometry as? SCNPlane else { return }

        // Update plane extent
        plane.width = CGFloat(planeAnchor.extent.x)
        plane.height = CGFloat(planeAnchor.extent.z)
    }

    func renderer(_ renderer: SCNSceneRenderer, didRemove node: SCNNode, for anchor: ARAnchor) {
        guard let planeAnchor = anchor as? ARPlaneAnchor else { return }
        planeOverlayNodes.removeValue(forKey: planeAnchor as ARAnchor)
    }

    private func hidePlaneOverlays() {
        for (_, planeNode) in planeOverlayNodes {
            planeNode.isHidden = true
        }
    }

    private func showPlaneOverlays() {
        hasPlacedFirstModel = false
        for (_, planeNode) in planeOverlayNodes {
            planeNode.isHidden = false
        }
    }

    // MARK: - ARSessionDelegate (Tracking State)

    func session(_ session: ARSession, cameraDidChangeTrackingState camera: ARCamera) {
        let newState: String
        switch camera.trackingState {
        case .normal:
            newState = "normal"
        case .limited(let reason):
            newState = "limited"
            let reasonStr: String
            switch reason {
            case .excessiveMotion:
                reasonStr = "excessive_motion"
            case .insufficientFeatures:
                reasonStr = "insufficient_features"
            case .initializing:
                reasonStr = "initializing"
            case .relocalizing:
                reasonStr = "relocalizing"
            @unknown default:
                reasonStr = "unknown"
            }
            lastTrackingState = newState
            onTrackingStateChange(["state": newState, "reason": reasonStr])
            return
        case .notAvailable:
            newState = "unavailable"
        }

        if newState != lastTrackingState {
            lastTrackingState = newState
            onTrackingStateChange(["state": newState])
        }
    }

    func session(_ session: ARSession, didFailWithError error: Error) {
        onARError([
            "code": "AR_SESSION_FAILED",
            "message": error.localizedDescription
        ])
    }

    // MARK: - Tap-to-Place

    @objc private func handleTap(_ gesture: UITapGestureRecognizer) {
        let location = gesture.location(in: arSceneView)

        // First: hit test existing model nodes (for selecting active gesture node)
        let hitResults = arSceneView.hitTest(location, options: [
            SCNHitTestOption.searchMode: SCNHitTestSearchMode.all.rawValue
        ])

        for result in hitResults {
            let tappedNode = findModelRootNode(result.node)
            if let tappedNode = tappedNode, nodeConfigMap[tappedNode] != nil {
                activeGestureNode = tappedNode
                return
            }
        }

        // Second: no existing node tapped, try placing a pending model
        activeGestureNode = nil

        guard !pendingModelConfigs.isEmpty, currentPendingIndex < pendingModelConfigs.count else {
            return
        }

        // Raycast for plane placement
        guard let query = arSceneView.raycastQuery(
            from: location,
            allowing: .existingPlaneGeometry,
            alignment: .any
        ) else { return }

        let results = arSceneView.session.raycast(query)
        guard let firstResult = results.first else { return }

        // Create anchor at hit location
        let anchor = ARAnchor(transform: firstResult.worldTransform)
        arSceneView.session.add(anchor: anchor)

        let config = pendingModelConfigs[currentPendingIndex]
        currentPendingIndex += 1

        loadAndPlaceModel(anchor: anchor, config: config)
    }

    /// Walk up the node hierarchy to find the root model container node that's in nodeConfigMap
    private func findModelRootNode(_ node: SCNNode) -> SCNNode? {
        var current: SCNNode? = node
        while let n = current {
            if nodeConfigMap[n] != nil {
                return n
            }
            current = n.parent
        }
        return nil
    }

    // MARK: - Model Loading

    private func loadAndPlaceModel(anchor: ARAnchor, config: ModelConfig) {
        let modelId = config.id.isEmpty ? "\(config.sourceUri.hashValue)" : config.id

        Task {
            do {
                let localURL = try await resolveModelURL(config.sourceUri)
                let gltfAsset = try await loadGLTFAsset(from: localURL)
                let scene = try SCNScene(gltfAsset: gltfAsset)

                await MainActor.run {
                    // Create container node for the model
                    let containerNode = SCNNode()

                    // Clone all children from the GLTF scene root into our container
                    for child in scene.rootNode.childNodes {
                        containerNode.addChildNode(child.clone())
                    }

                    // Apply scale
                    let s = config.scale
                    containerNode.scale = SCNVector3(s, s, s)

                    // Apply rotation (euler angles in radians)
                    if config.rotation.count >= 3 {
                        containerNode.eulerAngles = SCNVector3(
                            config.rotation[0],
                            config.rotation[1],
                            config.rotation[2]
                        )
                    }

                    // Find the anchor node in the scene and attach the model
                    if let anchorNode = self.arSceneView.node(for: anchor) {
                        anchorNode.addChildNode(containerNode)
                    }

                    // Track placement
                    self.nodeConfigMap[containerNode] = config
                    self.placedAnchors.append(AnchorNodePair(anchor: anchor, node: containerNode))
                    self.activeGestureNode = containerNode

                    // Hide plane overlays after first placement
                    if !self.hasPlacedFirstModel {
                        self.hasPlacedFirstModel = true
                        self.hidePlaneOverlays()
                    }

                    self.onModelLoaded(["modelId": modelId])
                    self.onModelPlaced([
                        "modelId": modelId,
                        "anchorId": anchor.identifier.uuidString
                    ])
                }
            } catch {
                await MainActor.run {
                    self.onModelError([
                        "modelId": modelId,
                        "code": "MODEL_LOAD_FAILED",
                        "message": error.localizedDescription
                    ])
                }
            }
        }
    }

    /// Resolve a model URI to a local file URL.
    /// Supports "asset://..." (bundled assets) and "https://..." (remote download).
    private func resolveModelURL(_ uri: String) async throws -> URL {
        if uri.hasPrefix("asset://") {
            // Strip the "asset://" prefix
            let assetName = String(uri.dropFirst("asset://".count))
            // Try to find in the main bundle
            if let url = Bundle.main.url(forResource: assetName, withExtension: nil) {
                return url
            }
            // Try without extension
            let name = (assetName as NSString).deletingPathExtension
            let ext = (assetName as NSString).pathExtension
            if let url = Bundle.main.url(forResource: name, withExtension: ext.isEmpty ? nil : ext) {
                return url
            }
            throw ARViewError.assetNotFound(assetName)
        } else if uri.hasPrefix("http://") || uri.hasPrefix("https://") {
            guard let remoteURL = URL(string: uri) else {
                throw ARViewError.invalidURL(uri)
            }
            return try await downloadFile(from: remoteURL)
        } else {
            // Assume it's a local file path
            let url = URL(fileURLWithPath: uri)
            if FileManager.default.fileExists(atPath: url.path) {
                return url
            }
            throw ARViewError.assetNotFound(uri)
        }
    }

    /// Download a remote file to a temporary location
    private func downloadFile(from remoteURL: URL) async throws -> URL {
        let (tempURL, response) = try await URLSession.shared.download(from: remoteURL)

        let httpResponse = response as? HTTPURLResponse
        guard let statusCode = httpResponse?.statusCode,
              (200...299).contains(statusCode) else {
            let code = httpResponse?.statusCode ?? 0
            throw ARViewError.downloadFailed("\(remoteURL.absoluteString) (HTTP \(code))")
        }

        // Move to a temp location with the original file extension
        let ext = remoteURL.pathExtension
        let destURL = FileManager.default.temporaryDirectory
            .appendingPathComponent(UUID().uuidString)
            .appendingPathExtension(ext)

        try FileManager.default.moveItem(at: tempURL, to: destURL)
        return destURL
    }

    /// Load a GLTFAsset from a local file URL using GLTFKit2
    private func loadGLTFAsset(from url: URL) async throws -> GLTFAsset {
        return try await withCheckedThrowingContinuation { continuation in
            var hasResumed = false
            GLTFAsset.load(with: url, options: [:]) { (_, status, asset, error, _) in
                guard !hasResumed else { return }
                switch status {
                case .complete:
                    hasResumed = true
                    if let asset = asset {
                        continuation.resume(returning: asset)
                    } else {
                        continuation.resume(throwing: error ?? ARViewError.modelLoadFailed("Unknown error"))
                    }
                case .error:
                    hasResumed = true
                    continuation.resume(throwing: error ?? ARViewError.modelLoadFailed("GLTF load error"))
                default:
                    break
                }
            }
        }
    }

    // MARK: - Gesture: Pinch to Scale

    @objc private func handlePinch(_ gesture: UIPinchGestureRecognizer) {
        guard let node = activeGestureNode,
              let config = nodeConfigMap[node],
              config.gestureScale else { return }

        if gesture.state == .changed {
            let rawFactor = Float(gesture.scale)
            let sensitivity = config.gestureScaleSensitivity
            let dampenedFactor = 1.0 + (rawFactor - 1.0) * sensitivity

            let currentScale = node.scale.x
            var newScale = currentScale * dampenedFactor
            newScale = min(max(newScale, config.gestureScaleMin), config.gestureScaleMax)

            node.scale = SCNVector3(newScale, newScale, newScale)

            // Reset for incremental updates
            gesture.scale = 1.0
        }
    }

    // MARK: - Gesture: Rotation

    @objc private func handleRotation(_ gesture: UIRotationGestureRecognizer) {
        guard let node = activeGestureNode,
              let config = nodeConfigMap[node],
              config.gestureRotate else { return }

        if gesture.state == .changed {
            node.eulerAngles.y -= Float(gesture.rotation)
            gesture.rotation = 0
        }
    }

    // MARK: - Gesture: Pan to Drag

    @objc private func handlePan(_ gesture: UIPanGestureRecognizer) {
        guard let node = activeGestureNode,
              nodeConfigMap[node] != nil else { return }

        let location = gesture.location(in: arSceneView)

        guard let query = arSceneView.raycastQuery(
            from: location,
            allowing: .existingPlaneGeometry,
            alignment: .any
        ) else { return }

        let results = arSceneView.session.raycast(query)
        guard let result = results.first else { return }

        switch gesture.state {
        case .changed, .began:
            // Get the hit world position
            let hitWorldPosition = SCNVector3(
                result.worldTransform.columns.3.x,
                result.worldTransform.columns.3.y,
                result.worldTransform.columns.3.z
            )

            // Convert world position to local position relative to the anchor (parent) node
            if let parentNode = node.parent {
                let localPos = parentNode.convertPosition(hitWorldPosition, from: nil)
                node.position = localPos
            }
        default:
            break
        }
    }

    // MARK: - Scene Stack Management

    func setModels(_ models: [ModelConfig]) {
        pendingModelConfigs = models
        currentPendingIndex = 0
        currentModels = models

        // Re-enable plane visualization so user can see surfaces to tap
        if !models.isEmpty {
            showPlaneOverlays()
        }
    }

    func pushScene(_ models: [ModelConfig]) {
        // Save current state
        let state = ARSceneState(
            models: currentModels,
            anchorNodes: placedAnchors
        )
        sceneStack.append(state)

        // Hide current anchor nodes
        for pair in placedAnchors {
            pair.node.isHidden = true
        }

        placedAnchors = []
        setModels(models)

        onSceneChange(["action": "push", "depth": sceneStack.count])
    }

    func popScene() -> Bool {
        guard !sceneStack.isEmpty else { return false }

        // Destroy current anchor nodes
        for pair in placedAnchors {
            cleanupAnchorNodeConfigs(pair)
            arSceneView.session.remove(anchor: pair.anchor)
            pair.node.removeFromParentNode()
        }
        placedAnchors = []

        // Restore previous state
        let previous = sceneStack.removeLast()
        currentModels = previous.models
        pendingModelConfigs = []
        currentPendingIndex = 0

        for pair in previous.anchorNodes {
            pair.node.isHidden = false
            placedAnchors.append(pair)
        }

        onSceneChange(["action": "pop", "depth": sceneStack.count])
        return true
    }

    func replaceScene(_ models: [ModelConfig]) {
        // Destroy current anchor nodes
        for pair in placedAnchors {
            cleanupAnchorNodeConfigs(pair)
            arSceneView.session.remove(anchor: pair.anchor)
            pair.node.removeFromParentNode()
        }
        placedAnchors = []

        setModels(models)
        onSceneChange(["action": "replace", "depth": sceneStack.count])
    }

    func popToTop() {
        guard !sceneStack.isEmpty else { return }

        // Destroy current anchor nodes
        for pair in placedAnchors {
            cleanupAnchorNodeConfigs(pair)
            arSceneView.session.remove(anchor: pair.anchor)
            pair.node.removeFromParentNode()
        }
        placedAnchors = []

        // Destroy all intermediate states
        while sceneStack.count > 1 {
            let state = sceneStack.removeLast()
            for pair in state.anchorNodes {
                cleanupAnchorNodeConfigs(pair)
                arSceneView.session.remove(anchor: pair.anchor)
                pair.node.removeFromParentNode()
            }
        }

        // Restore the first (bottom) state
        let first = sceneStack.removeLast()
        currentModels = first.models
        pendingModelConfigs = []
        currentPendingIndex = 0

        for pair in first.anchorNodes {
            pair.node.isHidden = false
            placedAnchors.append(pair)
        }

        onSceneChange(["action": "popToTop", "depth": 0])
    }

    /// Remove nodeConfigMap entries for a given anchor-node pair
    private func cleanupAnchorNodeConfigs(_ pair: AnchorNodePair) {
        nodeConfigMap.removeValue(forKey: pair.node)
        for child in pair.node.childNodes {
            nodeConfigMap.removeValue(forKey: child)
        }
    }
}

// MARK: - Error Types

enum ARViewError: LocalizedError {
    case assetNotFound(String)
    case invalidURL(String)
    case downloadFailed(String)
    case modelLoadFailed(String)

    var errorDescription: String? {
        switch self {
        case .assetNotFound(let name):
            return "Asset not found: \(name)"
        case .invalidURL(let url):
            return "Invalid URL: \(url)"
        case .downloadFailed(let url):
            return "Failed to download: \(url)"
        case .modelLoadFailed(let reason):
            return "Failed to load model: \(reason)"
        }
    }
}

File: android/src/main/java/expo/modules/arview/ReactNativeArView.kt
package expo.modules.arview

import android.content.Context
import android.util.Log
import android.view.MotionEvent
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.LifecycleRegistry
import com.google.ar.core.Config
import com.google.ar.core.Frame
import com.google.ar.core.Plane
import com.google.ar.core.TrackingFailureReason
import com.google.ar.core.TrackingState as ARTrackingState
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.viewevent.EventDispatcher
import expo.modules.kotlin.views.ExpoView
import io.github.sceneview.ar.ARSceneView
import io.github.sceneview.ar.node.AnchorNode
import io.github.sceneview.node.ModelNode
import io.github.sceneview.gesture.GestureDetector
import io.github.sceneview.node.Node
import android.view.ScaleGestureDetector as AndroidScaleDetector
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import java.lang.ref.WeakReference

class ReactNativeArView(context: Context, appContext: AppContext) :
    ExpoView(context, appContext), LifecycleOwner {

    companion object {
        private const val TAG = "ReactNativeArView"
        // Static weak reference for module access
        var currentInstance: WeakReference<ReactNativeArView>? = null
    }

    // Events
    private val onTrackingStateChange by EventDispatcher()
    private val onPlaneDetected by EventDispatcher()
    private val onModelLoaded by EventDispatcher()
    private val onModelPlaced by EventDispatcher()
    private val onModelError by EventDispatcher()
    private val onSceneChange by EventDispatcher()
    private val onARError by EventDispatcher()

    // Lifecycle
    private val lifecycleRegistry = LifecycleRegistry(this)
    override val lifecycle: Lifecycle get() = lifecycleRegistry

    // Scene management
    private val sceneStack = ArrayDeque<ARSceneState>()
    private var currentModels: MutableList<ModelConfig> = mutableListOf()
    private val placedAnchors: MutableList<AnchorNode> = mutableListOf()
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    // Pending models waiting for tap-to-place
    private var pendingModelConfigs: MutableList<ModelConfig> = mutableListOf()
    private var currentPendingIndex = 0

    // Map model nodes to their configs for dampened scale gestures
    private val nodeConfigMap = mutableMapOf<ModelNode, ModelConfig>()

    // Currently active model node for scale gesture
    private var activeScaleNode: ModelNode? = null

    // Tracking
    private var lastTrackingState: String = "unavailable"
    private val detectedPlaneIds = mutableSetOf<String>()

    // Plane detection mode
    private var planeFindingMode = Config.PlaneFindingMode.HORIZONTAL_AND_VERTICAL

    // AR view
    internal val arSceneView: ARSceneView

    // Current frame reference for hit testing
    private var currentFrame: Frame? = null

    fun setPlaneDetection(mode: String) {
        planeFindingMode = when (mode) {
            "horizontal" -> Config.PlaneFindingMode.HORIZONTAL
            "vertical" -> Config.PlaneFindingMode.VERTICAL
            else -> Config.PlaneFindingMode.HORIZONTAL_AND_VERTICAL
        }
        // Reconfigure the session if already running
        arSceneView.session?.let { session ->
            val config = session.config.apply {
                planeFindingMode = this@ReactNativeArView.planeFindingMode
            }
            session.configure(config)
        }
    }

    init {
        currentInstance = WeakReference(this)
        lifecycleRegistry.currentState = Lifecycle.State.CREATED

        arSceneView = try {
            ARSceneView(
                context = context,
                sessionConfiguration = { session, config ->
                    config.depthMode = when (session.isDepthModeSupported(Config.DepthMode.AUTOMATIC)) {
                        true -> Config.DepthMode.AUTOMATIC
                        else -> Config.DepthMode.DISABLED
                    }
                    config.lightEstimationMode = Config.LightEstimationMode.ENVIRONMENTAL_HDR
                    config.planeFindingMode = planeFindingMode
                },
                onSessionUpdated = { _, frame ->
                    handleFrameUpdate(frame)
                },
                onTrackingFailureChanged = { reason ->
                    handleTrackingFailure(reason)
                }
            ).apply {
                layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
                planeRenderer.isEnabled = true
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to create ARSceneView", e)
            onARError(mapOf(
                "code" to "AR_VIEW_INIT_FAILED",
                "message" to (e.message ?: "Failed to initialize AR view")
            ))
            // Create a minimal ARSceneView so the field is initialized
            ARSceneView(context).apply {
                layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
            }
        }

        addView(arSceneView)

        arSceneView.onGestureListener = object : GestureDetector.SimpleOnGestureListener() {
            override fun onSingleTapConfirmed(e: MotionEvent, node: Node?) {
                if (node == null) {
                    activeScaleNode = null
                    handleTapToPlace(e)
                } else {
                    activeScaleNode = (node as? ModelNode) ?: (node.parent as? ModelNode)
                }
            }
        }
    }

    // Our own scale detector - bypasses SceneView's built-in scale entirely
    private val scaleDetector = AndroidScaleDetector(context,
        object : AndroidScaleDetector.SimpleOnScaleGestureListener() {
            override fun onScale(detector: AndroidScaleDetector): Boolean {
                val node = activeScaleNode ?: return false
                val config = nodeConfigMap[node] ?: return false
                if (!config.gestureScale) return false

                val rawFactor = detector.scaleFactor
                val sensitivity = config.gestureScaleSensitivity
                val dampenedFactor = 1f + (rawFactor - 1f) * sensitivity

                val currentScale = node.scale.x
                val newScale = (currentScale * dampenedFactor)
                    .coerceIn(config.gestureScaleMin..config.gestureScaleMax)
                node.scale = dev.romainguy.kotlin.math.Float3(newScale, newScale, newScale)
                return true
            }
        }
    )

    override fun dispatchTouchEvent(event: MotionEvent): Boolean {
        scaleDetector.onTouchEvent(event)
        return super.dispatchTouchEvent(event)
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        lifecycleRegistry.currentState = Lifecycle.State.RESUMED
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        lifecycleRegistry.currentState = Lifecycle.State.DESTROYED
        scope.cancel()
        if (currentInstance?.get() === this) {
            currentInstance = null
        }
    }

    // --- Frame updates ---

    private fun handleFrameUpdate(frame: Frame) {
        currentFrame = frame

        val camera = frame.camera
        val newState = when (camera.trackingState) {
            ARTrackingState.TRACKING -> "normal"
            ARTrackingState.PAUSED -> "limited"
            ARTrackingState.STOPPED -> "unavailable"
        }
        if (newState != lastTrackingState) {
            lastTrackingState = newState
            onTrackingStateChange(mapOf("state" to newState))
        }

        for (plane in frame.getUpdatedTrackables(Plane::class.java)) {
            if (plane.trackingState == ARTrackingState.TRACKING) {
                val planeId = plane.hashCode().toString()
                if (planeId !in detectedPlaneIds) {
                    detectedPlaneIds.add(planeId)
                    val type = when (plane.type) {
                        Plane.Type.HORIZONTAL_UPWARD_FACING -> "horizontal"
                        Plane.Type.HORIZONTAL_DOWNWARD_FACING -> "horizontal"
                        Plane.Type.VERTICAL -> "vertical"
                    }
                    onPlaneDetected(mapOf("id" to planeId, "type" to type))
                }
            }
        }
    }

    private fun handleTrackingFailure(reason: TrackingFailureReason?) {
        if (reason == null || reason == TrackingFailureReason.NONE) return
        val reasonStr = when (reason) {
            TrackingFailureReason.BAD_STATE -> "bad_state"
            TrackingFailureReason.INSUFFICIENT_LIGHT -> "insufficient_light"
            TrackingFailureReason.EXCESSIVE_MOTION -> "excessive_motion"
            TrackingFailureReason.INSUFFICIENT_FEATURES -> "insufficient_features"
            TrackingFailureReason.CAMERA_UNAVAILABLE -> "camera_unavailable"
            else -> "unknown"
        }
        onTrackingStateChange(mapOf("state" to "limited", "reason" to reasonStr))
    }

    // --- Tap to place ---

    fun handleTapToPlace(event: MotionEvent) {
        if (pendingModelConfigs.isEmpty() || currentPendingIndex >= pendingModelConfigs.size) return
        val frame = currentFrame ?: return

        val hitResults = frame.hitTest(event.x, event.y)
        val validHit = hitResults.firstOrNull { hit ->
            val trackable = hit.trackable
            trackable is Plane && trackable.isPoseInPolygon(hit.hitPose) &&
                trackable.trackingState == ARTrackingState.TRACKING
        } ?: return

        val anchor = try {
            validHit.createAnchor()
        } catch (e: Exception) {
            Log.e(TAG, "Failed to create anchor", e)
            onARError(mapOf(
                "code" to "ANCHOR_CREATE_FAILED",
                "message" to (e.message ?: "Failed to create anchor")
            ))
            return
        }
        val modelConfig = pendingModelConfigs[currentPendingIndex]
        currentPendingIndex++

        loadAndPlaceModel(anchor, modelConfig)
    }

    private fun loadAndPlaceModel(anchor: com.google.ar.core.Anchor, config: ModelConfig) {
        val anchorNode = AnchorNode(arSceneView.engine, anchor)
        val modelId = config.id.ifEmpty { config.sourceUri.hashCode().toString() }

        scope.launch {
            try {
                val modelInstance = arSceneView.modelLoader.loadModelInstance(config.sourceUri)

                if (modelInstance != null) {
                    val modelNode = ModelNode(
                        modelInstance = modelInstance,
                        scaleToUnits = config.scale
                    ).apply {
                        // Scale handled by our own ScaleGestureDetector with dampening
                        isEditable = config.gestureRotate
                        isScaleEditable = false
                        isRotationEditable = config.gestureRotate
                        if (config.rotation.any { it != 0f }) {
                            rotation = dev.romainguy.kotlin.math.Float3(
                                config.rotation[0],
                                config.rotation[1],
                                config.rotation[2]
                            )
                        }
                    }

                    nodeConfigMap[modelNode] = config
                    activeScaleNode = modelNode
                    anchorNode.addChildNode(modelNode)
                    arSceneView.addChildNode(anchorNode)
                    placedAnchors.add(anchorNode)

                    // Hide plane dots once a model is placed
                    arSceneView.planeRenderer.isEnabled = false

                    onModelLoaded(mapOf("modelId" to modelId))
                    onModelPlaced(mapOf(
                        "modelId" to modelId,
                        "anchorId" to anchor.hashCode().toString()
                    ))
                } else {
                    onModelError(mapOf(
                        "modelId" to modelId,
                        "code" to "MODEL_LOAD_FAILED",
                        "message" to "Failed to load model from ${config.sourceUri}"
                    ))
                }
            } catch (e: Exception) {
                Log.e(TAG, "Failed to load model", e)
                onModelError(mapOf(
                    "modelId" to modelId,
                    "code" to "MODEL_LOAD_FAILED",
                    "message" to (e.message ?: "Unknown error loading model")
                ))
            }
        }
    }

    private fun cleanupAnchorNodeConfigs(anchorNode: AnchorNode) {
        for (child in anchorNode.childNodes) {
            (child as? ModelNode)?.let { nodeConfigMap.remove(it) }
        }
    }

    // --- Scene stack management ---

    fun setModels(models: List<ModelConfig>) {
        pendingModelConfigs = models.toMutableList()
        currentPendingIndex = 0
        currentModels = models.toMutableList()
        // Re-enable plane renderer so user can see surfaces to tap
        if (models.isNotEmpty()) {
            arSceneView.planeRenderer.isEnabled = true
        }
    }

    fun pushScene(models: List<ModelConfig>) {
        sceneStack.addLast(ARSceneState(
            models = currentModels.toList(),
            anchorNodes = placedAnchors.toList()
        ))

        for (node in placedAnchors) {
            node.isVisible = false
        }

        placedAnchors.clear()
        setModels(models)

        onSceneChange(mapOf("action" to "push", "depth" to sceneStack.size))
    }

    fun popScene(): Boolean {
        if (sceneStack.isEmpty()) return false

        for (node in placedAnchors) {
            cleanupAnchorNodeConfigs(node)
            arSceneView.removeChildNode(node)
            node.destroy()
        }
        placedAnchors.clear()

        val previous = sceneStack.removeLast()
        currentModels = previous.models.toMutableList()
        pendingModelConfigs = mutableListOf()
        currentPendingIndex = 0

        for (node in previous.anchorNodes) {
            node.isVisible = true
            placedAnchors.add(node)
        }

        onSceneChange(mapOf("action" to "pop", "depth" to sceneStack.size))
        return true
    }

    fun replaceScene(models: List<ModelConfig>) {
        for (node in placedAnchors) {
            cleanupAnchorNodeConfigs(node)
            arSceneView.removeChildNode(node)
            node.destroy()
        }
        placedAnchors.clear()
        setModels(models)
        onSceneChange(mapOf("action" to "replace", "depth" to sceneStack.size))
    }

    fun popToTop() {
        if (sceneStack.isEmpty()) return

        for (node in placedAnchors) {
            cleanupAnchorNodeConfigs(node)
            arSceneView.removeChildNode(node)
            node.destroy()
        }
        placedAnchors.clear()

        while (sceneStack.size > 1) {
            val state = sceneStack.removeLast()
            for (node in state.anchorNodes) {
                cleanupAnchorNodeConfigs(node)
                arSceneView.removeChildNode(node)
                node.destroy()
            }
        }

        val first = sceneStack.removeLast()
        currentModels = first.models.toMutableList()
        pendingModelConfigs = mutableListOf()
        currentPendingIndex = 0

        for (node in first.anchorNodes) {
            node.isVisible = true
            placedAnchors.add(node)
        }

        onSceneChange(mapOf("action" to "popToTop", "depth" to 0))
    }
}
